pycloudlib/azure/cloud.py

```diff
@@ -0,0 +1,644 @@
+            "ip_address": ip_address.ip_address,
+            "rg_name": self.resource_group.name
+        }
+
+        instance = AzureInstance(
+            client=self.compute_client,
+            key_pair=self.key_pair,
+            instance=instance_info
+        )
+
+        if wait:
+            instance.wait()
+
+        self.registered_instances[vm.name] = instance
+        return instance
+
+    def _create_ssh_resource(self, key_name):
+        """Create a ssh resource.
+
+        This method creates an Azure ssh resource to be associated
+        with a resource group.
+
+        Args:
+            key_name: string, The name of the ssh resource.
+
+        """
+        self.compute_client.ssh_public_keys.create(
+            self.resource_group.name,
+            key_name,
+            parameters={
+                "location": self.location
+            }
+        )
+
+    def create_key_pair(self, key_name):
+        """Create a pair of ssh keys.
+
+        This method creates an a pair of ssh keys in
+        the class resource group.
+
+        Args:
+            key_name: string, The name of the ssh resource.
+
+        """
+        self._create_ssh_resource(key_name)
+
+        ssh_call = self.compute_client.ssh_public_keys.generate_key_pair(
+            resource_group_name=self.resource_group.name,
+            ssh_public_key_name=key_name)
+
+        return ssh_call.public_key, ssh_call.private_key
+
+    def list_keys(self):
+        """List all ssh keys in the class resource group."""
+        ssh_public_keys = self.compute_client.ssh_public_keys
+
+        return [
+            ssh.name
+            for ssh in ssh_public_keys.list_by_resource_group(
+                self.resource_group.name)
+        ]
+
+    def delete_key(self, key_name):
+        """Delete a ssh key from the class resource group.
+
+        Args:
+            key_name: string, The name of the ssh resource.
+
+        """
+        ssh_public_keys = self.compute_client.ssh_public_keys
+        ssh_public_keys.delete(
+            resource_group_name=self.resource_group.name,
+            ssh_public_key_name=key_name
+        )
+
+    def use_key(self, public_key_path, private_key_path=None, name=None):
+        """Use an existing already uploaded key.
+
+        Args:
+            public_key_path: path to the public key to upload
+            private_key_path: path to the private key to upload
+            name: name to reference key by
+
+        """
+        if not name:
+            name = self.tag
+
+        self._log.debug('using SSH key %s', name)
+        self.key_pair = KeyPair(public_key_path, private_key_path, name)
+
+    def get_instance(self, instance_id):
+        """Get an instance by id.
+
+        Args:
+            instance_id: the instance name to search by
+
+        Returns:
+            An instance object to use to manipulate the instance further.
+
+        """
+        if instance_id in self.registered_instances:
```

Comment:
This really needs to be calling azure.compute_client.virtual_machines.list or .get(vm_name='blah') api.
What this may mean is that we have to either provide an optional resource_group parameter to search for the named instance in your desired groupname